quinn-proto/src/connection.rs

```diff
@@ -765,13 +765,32 @@ impl Connection {
         remote: SocketAddrV6,
         partial_decode: PartialDecode,
     ) -> Option<BytesMut> {
-        let result = {
-            let crypto = if partial_decode.is_handshake() {
-                &self.handshake_crypto
+        let result = if partial_decode.is_handshake() {
+            partial_decode.finish(self.handshake_crypto.pn_decrypt_key())
+        } else if partial_decode.key_phase() != self.key_phase {
+            let new = self.crypto.as_mut().unwrap().update();
+            let decoded = partial_decode.finish(new.pn_decrypt_key());
+
+            if let Ok((ref packet, ..)) = decoded {
+                let number = match packet.header {
+                    Header::Short { number, .. } => number,
+                    _ => {
+                        warn!(self.log, "unable to complete packet decoding with new key");
+                        return None;
+                    }
+                };
+
+                let decoded_number = number.expand(self.rx_packet + 1);
+                let old = mem::replace(self.crypto.as_mut().unwrap(), new);
+                self.prev_crypto = Some((decoded_number, old));
+                self.key_phase = !self.key_phase;
             } else {
-                &self.crypto.as_ref().unwrap()
-            };
-            partial_decode.finish(crypto.pn_decrypt_key())
+                warn!(self.log, "unable to complete packet decoding with new key");
+                return None;
+            }
+            decoded
+        } else {
```

Comment:
Can we sort this simple case before the complex one? I think that would make it more readable.